### PR TITLE
Track motion events for reuse post gesture disambiguation

### PR DESCRIPTION
**Pull request by iskakaushik**
This change makes it so that we track all the motion events encountered by `FlutterView` and all of its subviews in the `MotionEventTracker` class, indexed by a unique `MotionEventId`. This identifier is then passed to the Flutter framework as seen in https://github.com/flutter/flutter/pull/60930. Once the gestures take part in gesture disambiguation and are sent back to the engine, we look-up the original motion event using the `MotionEventId` and dispatch it to the platform.

Bug: https://github.com/flutter/flutter/issues/58837

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java [hunk 1]
```diff
@@ -183,6 +187,7 @@ private void addPointerForIndex(
 
     long timeStamp = event.getEventTime() * 1000; // Convert from milliseconds to microseconds.
 
+    packet.putLong(motionEventId.getId());
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:190]: do you know if the other embeddings have to be updated as well? I wonder if adding the field to the end is safer.

File: shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+/** Tracks the motion events received by the FlutterView. */
+public final class MotionEventTracker {
+
+  /** Represents a unique identifier corresponding to a motion event. */
+  public static class MotionEventId {
+    private static final AtomicLong ID_COUNTER = new AtomicLong(0);
+    private final long id;
+
+    private MotionEventId(long id) {
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java:17]: would this wrapper object provide something that just the return value of `AtomicLong` doesn't?

File: shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * Returns the MotionEvent corresponding to the eventId while discarding all the motion events
+   * that occured prior to the event represented by the eventId. Returns null if this event was
+   * popped or discarded.
+   */
+  @Nullable
+  public MotionEvent pop(MotionEventId eventId) {
+    // TODO(kaushikiska) do the actual timestamp based book-keeping.
+    return eventById.remove(eventId);
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java:81]: do you plan to remove all the previous events as well?  How would we make sure that the queue of gestures doesn't grow for ever?

For example:
If the motion events aren't owned by the Android view,  how would we make sure that those are ignored or removed?

If there are N motion events, but only the last one is owned by the Android view, where would we delete the motion events owned by Flutter widgets?

Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java:81]: SGTM :)

File: shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java (hunk 2 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/MotionEventTracker.java:17]: Got it. I don't feel strongly about it, so your call!

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java (hunk 1 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:190]: SG

**Comment by xster** [comment]
(go/tested-engine-prs nag)
This PR could use a test

**Pull request merged**